draft-ietf-quic-tls.md

```diff
@@ -866,6 +867,85 @@ the connection that is hosted on stream 0.  This sequence number is not visible
 to QUIC.
 
 
+## Packet Number Protection {#pn-encrypt}
+
+QUIC packets are protected using a key that is derived from the current set of
+secrets.  The key derived using the "pn" label is used to protect the packet
+number from casual observation.  The packet number protection algorithm depends
+on the negotiated AEAD.
+
+Packet number protection is applied after packet protection is applied (see
+{{aead}}).  The ciphertext of the packet is sampled and used as input to an
+encryption algorithm.  For packets with a long header, the ciphertext starting
+immediately after the packet number is used (that is, octet 17 onwards).  For
+packets with a short header, the packet number length is not known, so it is
+assumed to be its largest possible length (4 octets).  Thus, for a short header,
+the sampled ciphertext starts at either octet 5 when the connection ID is
+omitted, or octet 13 when the connection is present.
+
```

Comment:
If packet number protection (for good reasons) is only applied after handshake, this prevents connection migration from looking like a new handshake unless there are special rules for packet numbering during a handshake. Hence, perhaps protetected packet numbers should have a separate numbering space.

Reply:
As it stands now, connection migration in no way looks like a new connection, because no long form handshake packets are used on the new path.

Reply:
> As it stands now,

I was thinking a bit ahead - considering the open issue(s?) on the topic.

Reply:
Perhaps a client-initiated PATH_CHALLENGE/PATH_RESPONSE should be long header packets?  But that's totally orthogonal to this PR, I think.  So long as the packet number is obscured in both header types, an observer can't tell whether a connection is restarting from zero or not.

Reply:
The initial handshake needs to negotiate encryption and hence cannot encrypt those packets.

Perhaps the packet numbers ought to be encrypted with a version specific non-negotiably algorithm, but that hurts devices that doesn't speak AES natively. This is why I suggest each handshake and migration sequence uses unencrypted, from 0, packet numbers, unique to each connection ID.

Reply:
The biggest problem for having migration appear to be a new connection is that the packets won't decrypt with the handshake keys.  This doesn't change any of that.  @mikkelfj separately suggested that we use 0-RTT, and that could work, but it's a pretty big lift.

Reply:
Yes that is true - in 0-RTT the packet numbers could be encrypted.

BTW: if 0-RTT doesn't actively take measures to avoid fingerprinting, that would be an excellent way to track users. This somewhat relates to the replay thing.